### ISSUE TITLE
BUG: Inconsistent `deriv` definition in maps

### ISSUE THREAD

**Issue by jcapriot** [simpeg/simpeg#1389]
### Describe the issue:

Not all maps follow the signature of the base map `IdentityMap`,
It should be:
```python
def deriv(self, m, v=None):
    ...
```

The maps that currently do not accept `v`:
- [ ] `SelfConsistentEffectiveMedium`
- [ ] `ParametricLayer`
- [ ] `ParametricBlock`
- [ ] `ParametricEllipsoid`
- [ ] `ParametricCasingAndLayer`
- [ ] `ParametricBlockInLayer`

### Reproducible code example:

```python
from SimPEG import maps
import numpy as np
import discretize

mesh = discretize.TensorMesh([np.ones(10)], "C")
block = maps.ParametricBlock(mesh)

model = np.ones(mesh.n_cells)

block.deriv(model, v=None)
```

### Error message:

```shell
ParametricBlock.deriv() got an unexpected keyword argument 'v'
```

### Runtime information:

...

**Comment by YingHuuu** [comment]
Gi, Joe! May I start with ```ParametricLayer``` and try to handle it?

**Comment by santisoler** [comment]
Today we were discussing with @YingHuuu what to do with the `v` arguments in these methods and agreed that the `deriv` method should return the matrix multiplied by the passed vector. For example:

```python
def deriv(self, m, v=None):
    # Compute the matrix    
    matrix = ...
    if v is not None:
         return matrix @ v
    return matrix
```

Moreover, we should also check that the docstrings agree with what the methods are doing. We should:
- mark `v` as optional,
- check that the return specifies `scipy.sparse.csr_matrix or numpy.ndarray`.

For example:
```
        Parameters
        ----------
        m : (nP) numpy.ndarray
            A vector representing a set of model parameters
        v : (nP) numpy.ndarray, optional
            If not ``None``, the method returns the derivative times the vector ``v``.

        Returns
        -------
        scipy.sparse.csr_matrix or numpy.ndarray
            Derivative of the mapping with respect to the model parameters. 
            If the input argument *v* is not ``None``, the method returns the
            derivative times the vector *v*.
```

**Comment by jcapriot** [comment]
> Today we were discussing with @YingHuuu what to do with the `v` arguments in these methods and agreed that the `deriv` method should return the matrix multiplied by the passed vector. For example:
> 
> ```python
> def deriv(self, m, v=None):
>     # Compute the matrix    
>     matrix = ...
>     if v is not None:
>          return matrix @ v
>     return matrix
> ```
> 
> Moreover, we should also check that the docstrings agree with what the methods are doing. We should:
> 
> * mark `v` as optional,
> * check that the return specifies `scipy.sparse.csr_matrix or numpy.ndarray`.
> 
> For example:
> 
> ```
>         Parameters
>         ----------
>         m : (nP) numpy.ndarray
>             A vector representing a set of model parameters
>         v : (nP) numpy.ndarray, optional
>             If not ``None``, the method returns the derivative times the vector ``v``.
> 
>         Returns
>         -------
>         scipy.sparse.csr_matrix or numpy.ndarray
>             Derivative of the mapping with respect to the model parameters. 
>             If the input argument *v* is not ``None``, the method returns the
>             derivative times the vector *v*.
> ```

If possible, we should avoid constructing the sparse matrix on the call to `deriv` when `v is not None`.

**Comment by jcapriot** [comment]
Also, the requirement that the output is a `csr_matrix` when `v is None` is too strict. It should only be required to return a 2D array_like object. Basically anything that supports doing matrix multiplication via the `@` operator, and taking a transpose via `.T`. For example see how the `ComplexMap` returns a LinearOperator so that it can have make the transpose operator be the hermitian operator: 

https://github.com/simpeg/simpeg/blob/80e3b1807d8dedac04cb8aa1731957462ffa187c/simpeg/maps.py#L2648

**Comment by santisoler** [comment]
> Also, the requirement that the output is a `csr_matrix` when `v is None` is too strict. It should only be required to return a 2D array_like object. Basically anything that supports doing matrix multiplication via the `@` operator, and taking a transpose via `.T`. For example see how the `ComplexMap` returns a LinearOperator so that it can make the transpose operator be the hermitian operator:
> 
> https://github.com/simpeg/simpeg/blob/80e3b1807d8dedac04cb8aa1731957462ffa187c/simpeg/maps.py#L2648

Right. I think I wanted to make the point of double checking the docstrings match the outputs of the class. For the particular case that @YingHuuu is working now, the `csr_matrix` looks good enough.

**Comment by santisoler** [comment]
> If possible, we should avoid constructing the sparse matrix on the call to `deriv` when `v is not None`.

Good point! I think we can check case by case if that's possible.
